Problem resolutions are dissociated from their description in the build failure message

### Current Behavior

The Problems API allows clients to fail the build by throwing an exception with the `ProblemReporter.throwing(..)` methods. 

The build failure message has a specific rendering: the first part renders the description, then separately, there's a dedicated resolution section (ie the `Try:` section) containing the proposed solutions. When the build failure has associated Problems the solutions property from the Problems are also rendered in the resolution section. 

The problem arises when there's more than one Problem associated with the build failure. The solutions are dissociated from the descriptions. See the example build failure message:
```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/donat/Development/git/gradle/problems-playground/build.gradle.kts' line: 346

* What went wrong:
java.lang.Exception: Boom failure
> Problem 1
    Context for Problem 1
      Details about Problem 1
  Problem 2
    Context for Problem 2
      Details about Problem 2

* Try:
> Solution A
> Solution B
> Run with --scan to generate a Build Scan (powered by Develocity).

BUILD FAILED in 401ms
```

Which solution belongs to Problem 1 and to Problem 2?


### Expected Behavior

There are multiple ways we can fix this problem. One option is simply to get rid of the separation of problems and solutions and render them as a single unit. That is reasonable, however we have to revisit the existing failure template and figure out what to do with the generic advice under the Try section. 

Alternatively, the failure message could i) number the rendered problems and use those as references in the Try  section, ii) inline the solutions into the problem rendering, or something else. The numbering-based approach seems more streamlined: it would not result in a different format in different scenarios (0, 1, or more problem occurrences associated with a build failure). Also numbers can be omitted when there’s exactly one associated problem occurrence, so most build failure messages would be kept unchanged.

--- a/platforms/core-configuration/base-diagnostics/src/testFixtures/groovy/org/gradle/api/reporting/components/AbstractComponentReportIntegrationTest.groovy
+++ b/platforms/core-configuration/base-diagnostics/src/testFixtures/groovy/org/gradle/api/reporting/components/AbstractComponentReportIntegrationTest.groovy
@@ -43,7 +43,8 @@ abstract class AbstractComponentReportIntegrationTest extends AbstractIntegratio
             !it.isEmpty() && !(it ==~ /^Download http.*$/) && !(it ==~ /.*has been deprecated.*$/)
                 && !(it ==~ /.*Problem found.*$/)
                 && !(it ==~ /.*invocation of 'Task.project' at execution time is unsupported.*$/)
-                && !(it ==~ /.*Documentation: https:\/\/docs.gradle.org\/.*$/)
+                && !(it ==~ /.*Possible solution:.*$/)
+                && !(it ==~ /.*For more information, please refer to https:\/\/docs.gradle.org\/.*$/)
         }.join('\n')
     }
 

--- a/platforms/core-runtime/build-configuration/src/integTest/groovy/org/gradle/interal/buildconfiguration/tasks/UpdateDaemonJvmIntegrationTest.groovy
+++ b/platforms/core-runtime/build-configuration/src/integTest/groovy/org/gradle/interal/buildconfiguration/tasks/UpdateDaemonJvmIntegrationTest.groovy
@@ -21,6 +21,7 @@ import org.gradle.api.JavaVersion
 import org.gradle.buildconfiguration.tasks.UpdateDaemonJvm
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.executer.DocumentationUtils
 import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
 import org.gradle.integtests.fixtures.jvm.JavaToolchainFixture
 import org.gradle.internal.buildconfiguration.fixture.DaemonJvmPropertiesFixture
@@ -75,8 +76,8 @@ class UpdateDaemonJvmIntegrationTest extends AbstractIntegrationSpec implements 
         }
         failureHasCause('Invalid task configuration')
         failureCauseContains('Toolchain download repositories have not been configured.')
-        failure.assertHasResolution('Configure toolchain download repositories in your build settings.')
-        failure.assertHasResolution("For more information, see https://docs.gradle.org/current/userguide/toolchains.html#sub:download_repositories")
+        failure.error.contains('Possible solution: Configure toolchain download repositories in your build settings.')
+        failure.error.contains(DocumentationUtils.normalizeDocumentationLink('For more information, please refer to https://docs.gradle.org/current/userguide/toolchains.html#sub:download_repositories.'))
 
     }
 
@@ -333,8 +334,8 @@ tasks.named("updateDaemonJvm") {
 //        failureDescriptionContains("Execution failed for task ':updateDaemonJvm'")
         failureHasCause("Invalid task configuration")
         failureCauseContains("Toolchain resolvers did not return download URLs providing a JDK matching {languageVersion=20, vendor=vendor matching('FOO'), implementation=vendor-specific, nativeImageCapable=false} for any of the requested platforms")
-        failure.assertHasResolution('Use a toolchain download repository capable of resolving the toolchain spec for the given platforms.')
-        failure.assertHasResolution('For more information, see https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:daemon_jvm_provisioning')
+        failure.error.contains('Possible solution: Use a toolchain download repository capable of resolving the toolchain spec for the given platforms.')
+        failure.error.contains(DocumentationUtils.normalizeDocumentationLink('For more information, please refer to https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:daemon_jvm_provisioning.'))
     }
 
     @Requires(InstalledJdkTestPreconditions.JavaHomeWithDifferentVersionAvailable)

--- a/platforms/core-runtime/build-configuration/src/main/java/org/gradle/internal/buildconfiguration/DaemonJvmPropertiesConfigurator.java
+++ b/platforms/core-runtime/build-configuration/src/main/java/org/gradle/internal/buildconfiguration/DaemonJvmPropertiesConfigurator.java
@@ -86,11 +86,10 @@ public class DaemonJvmPropertiesConfigurator implements ProjectConfigureAction {
                             JavaToolchainResolverService resolverService = project.getServices().get(JavaToolchainResolverService.class);
                             if (!resolverService.hasConfiguredToolchainRepositories()) {
                                 UnconfiguredToolchainRepositoriesResolver exception = new UnconfiguredToolchainRepositoriesResolver();
-                                throw reporter.throwing(exception, UpdateDaemonJvm.TASK_CONFIGURATION_PROBLEM_ID,
-                                    problemSpec -> {
-                                        problemSpec.solution("Configure toolchain download repositories in your build settings.");
-                                        problemSpec.documentedAt(Documentation.userManual("toolchains", "sub:download_repositories").getUrl());
-                                });
+                                throw reporter.throwing(exception, reporter.create(UpdateDaemonJvm.TASK_CONFIGURATION_PROBLEM_ID, problemSpec -> {
+                                    problemSpec.solution("Configure toolchain download repositories in your build settings.");
+                                    problemSpec.documentedAt(Documentation.userManual("toolchains", "sub:download_repositories").getUrl());
+                                }));
                             }
                             Map<BuildPlatform, Optional<URI>> buildPlatformOptionalUriMap = platforms.stream()
                                 .collect(Collectors.toMap(platform -> platform,
@@ -99,12 +98,11 @@ public class DaemonJvmPropertiesConfigurator implements ProjectConfigureAction {
                                 .filter(e -> e.getValue().isPresent())
                                 .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().get()));
                             if (platformToDownloadUri.isEmpty()) {
-                                throw reporter.throwing(new IllegalStateException("Toolchain resolvers did not return download URLs providing a JDK matching " + toolchainSpec + " for any of the requested platforms " + platforms),
-                                    UpdateDaemonJvm.TASK_CONFIGURATION_PROBLEM_ID,
-                                    problemSpec -> {
-                                        problemSpec.solution("Use a toolchain download repository capable of resolving the toolchain spec for the given platforms.");
-                                        problemSpec.documentedAt(Documentation.userManual("gradle_daemon", "sec:daemon_jvm_provisioning").getUrl());
-                                    });
+                                IllegalStateException exception = new IllegalStateException("Toolchain resolvers did not return download URLs providing a JDK matching " + toolchainSpec + " for any of the requested platforms " + platforms);
+                                throw reporter.throwing(exception, reporter.create(UpdateDaemonJvm.TASK_CONFIGURATION_PROBLEM_ID, problemSpec -> {
+                                    problemSpec.solution("Use a toolchain download repository capable of resolving the toolchain spec for the given platforms.");
+                                    problemSpec.documentedAt(Documentation.userManual("gradle_daemon", "sec:daemon_jvm_provisioning").getUrl());
+                                }));
                             }
                             return platformToDownloadUri;
                         }));

--- a/platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy
+++ b/platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy
@@ -477,9 +477,9 @@ class ProblemsServiceIntegrationTest extends AbstractIntegrationSpec {
 Problem found: Project is a prototype (id: sample-problems:prototype-project)
   This is a prototype and not a guideline for modeling real-life projects
     Complex build logic like the Problems API usage should be integrated into plugins
-    Solution: Look up the samples index for real-life examples
-    Documentation: https://example.com/some-problem
-    Location: /path/to/script
+    For more information, please refer to https://example.com/some-problem.
+    Location: /path/to/script line 20
+    Possible solution: Look up the samples index for real-life examples.
         """
         verifyAll(receivedProblem) {
             definition.id.fqid == 'sample-problems:prototype-project'

--- a/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemBodyWriter.java
+++ b/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemBodyWriter.java
@@ -17,8 +17,10 @@
 package org.gradle.problems.internal.rendering;
 
 import com.google.common.base.Strings;
+import org.gradle.api.problems.DocLink;
 import org.gradle.api.problems.FileLocation;
 import org.gradle.api.problems.LineInFileLocation;
+import org.gradle.api.problems.internal.DocLinkInternal;
 import org.gradle.api.problems.internal.ProblemInternal;
 import org.gradle.util.internal.TextUtil;
 import org.jspecify.annotations.Nullable;
@@ -33,50 +35,73 @@ import java.util.stream.Collectors;
  */
 class ProblemBodyWriter implements PartialProblemWriter {
 
+    private static final int LEVEL_1_INDENT = 2;
+    private static final int LEVEL_2_INDENT = 4;
+    private static final int LEVEL_3_INDENT = 6;
+
     @Override
     public void write(ProblemInternal problem, RenderOptions options, PrintWriter output) {
         // contextual message, if any
         String problemSubMessage = getContextualMessage(problem);
         if (problemSubMessage != null) {
             output.printf("%n");
-            indent(output, problemSubMessage, 2);
+            indent(output, problemSubMessage, LEVEL_1_INDENT);
         }
 
         // indent details further if there was a contextual message
         if (problem.getDetails() != null) {
             output.printf("%n");
-            indent(output, problem.getDetails(), problemSubMessage == null ? 2 : 4);
+            indent(output, problem.getDetails(), problemSubMessage == null ? LEVEL_1_INDENT : LEVEL_2_INDENT);
         }
 
-        // print solutions
-        if (options.isRenderSolutions() && !problem.getSolutions().isEmpty()) {
-            for (String solution : problem.getSolutions()) {
-                output.printf("%n");
-
-                String[] lines = solution.split("\\r?\\n");
-                indent(output, "Solution: " + lines[0], 4);
-                for (int i = 1; i < lines.length; i++) {
-                    output.printf("%n");
-                    indent(output, lines[i], 14); // 4 + "Solution: ".length()
-                }
-            }
-        }
-
-        // print documentation
-        if (options.isRenderSolutions() && problem.getDefinition().getDocumentationLink() != null) {
+        // link to documentation
+        DocLink documentationLink = problem.getDefinition().getDocumentationLink();
+        if (documentationLink != null && documentationLink.getUrl() != null) {
             output.printf("%n");
-            indent(output, "Documentation: " + problem.getDefinition().getDocumentationLink().getUrl(), 4);
+            String message = documentationLink instanceof DocLinkInternal
+                ? ((DocLinkInternal) documentationLink).getConsultDocumentationMessage()
+                : String.format("For more information, please refer to %s.", documentationLink.getUrl());
+            indent(output, message, LEVEL_2_INDENT);
         }
 
-        // print locations
+        // locations
         List<FileLocation> fileLocations = problem.getOriginLocations().stream().filter(FileLocation.class::isInstance).map(FileLocation.class::cast).collect(Collectors.toList());
         for (FileLocation location : fileLocations) {
             output.printf("%n");
-            indent(output, "Location: " + location.getPath(), 4);
+            indent(output, "Location: " + location.getPath(), LEVEL_2_INDENT);
             if (location instanceof LineInFileLocation) {
                 LineInFileLocation lineLocation = (LineInFileLocation) location;
                 output.printf(" line " + lineLocation.getLine());
             }
+        }
+
+        // solutions
+        List<String> solutions = problem.getSolutions();
+        if (!solutions.isEmpty()) {
+            output.printf("%n");
+            if (solutions.size() == 1) {
+                writePrefixedMultiline(output, "Possible solution: ", LEVEL_2_INDENT, normalize(solutions.get(0)));
+            } else {
+                indent(output, "Possible solutions:", LEVEL_2_INDENT);
+                for (int i = 0; i < solutions.size(); i++) {
+                    output.printf("%n");
+                    writePrefixedMultiline(output, (i + 1) + ". ", LEVEL_3_INDENT, normalize(solutions.get(i)));
+                }
+            }
+        }
+    }
+
+    private static String normalize(String message) {
+        return TextUtil.capitalize(TextUtil.endLineWithDot(message));
+    }
+
+    private static void writePrefixedMultiline(PrintWriter output, String firstLinePrefix, int firstLineIndent, String text) {
+        String[] lines = text.split("\\r?\\n");
+        indent(output, firstLinePrefix + lines[0], firstLineIndent);
+        int continuationIndent = firstLineIndent + firstLinePrefix.length();
+        for (int i = 1; i < lines.length; i++) {
+            output.printf("%n");
+            indent(output, lines[i], continuationIndent);
         }
     }
 

--- a/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemWriter.java
+++ b/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemWriter.java
@@ -59,7 +59,7 @@ public abstract class ProblemWriter {
     public static ProblemWriter simple() {
         return new SimpleProblemWriter(
             ProblemWriterRegistry.INSTANCE,
-            new RenderOptions("Problem found: ", true, true)
+            new RenderOptions("Problem found: ", true)
         );
     }
 
@@ -70,7 +70,7 @@ public abstract class ProblemWriter {
     public static ProblemWriter grouping() {
         return new GroupingProblemWriter(
             ProblemWriterRegistry.INSTANCE,
-            new RenderOptions("", false, false));
+            new RenderOptions("", false));
     }
 
     /**

--- a/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/RenderOptions.java
+++ b/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/RenderOptions.java
@@ -23,16 +23,13 @@ class RenderOptions {
 
     private final String prefix;
     private final boolean renderId;
-    private final boolean renderSolutions;
 
     public RenderOptions(
         String prefix,
-        boolean renderId,
-        boolean renderSolutions
+        boolean renderId
     ) {
         this.prefix = prefix;
         this.renderId = renderId;
-        this.renderSolutions = renderSolutions;
     }
 
     /**
@@ -47,12 +44,5 @@ class RenderOptions {
      */
     public boolean isRenderId() {
         return renderId;
-    }
-
-    /**
-     * Whether to render the solutions with the problem.
-     */
-    public boolean isRenderSolutions() {
-        return renderSolutions;
     }
 }

--- a/platforms/ide/problems-rendering/src/test/groovy/org/gradle/problems/internal/rendering/GroupingProblemWriterTest.groovy
+++ b/platforms/ide/problems-rendering/src/test/groovy/org/gradle/problems/internal/rendering/GroupingProblemWriterTest.groovy
@@ -146,6 +146,71 @@ display-name
         ''')
     }
 
+    def "multiple problems with solutions and documentation link are properly separated"() {
+        given:
+        def problem1 = createProblemBuilder()
+            .id("id", "display-name", level1Group)
+            .contextualLabel("First context")
+            .details("First details")
+            .solution("First solution")
+            .documentedAt("https://example.com/first")
+            .build()
+        def problem2 = createProblemBuilder()
+            .id("id", "display-name", level1Group)
+            .contextualLabel("Second context")
+            .details("Second details")
+            .solution("Second solution A")
+            .solution("Second solution B")
+            .documentedAt("https://example.com/second")
+            .build()
+
+        when:
+        problemWriter.write([problem1, problem2], writer)
+
+        then:
+        renderedProblem == denormalizeAndStrip('''
+display-name
+  First context
+    First details
+    For more information, please refer to https://example.com/first.
+    Possible solution: First solution.
+display-name
+  Second context
+    Second details
+    For more information, please refer to https://example.com/second.
+    Possible solutions:
+      1. Second solution A.
+      2. Second solution B.
+        ''')
+    }
+
+    def "mixed problems with and without solutions are properly separated"() {
+        given:
+        def withSolution = createProblemBuilder()
+            .id("id", "display-name", level1Group)
+            .contextualLabel("With solution")
+            .solution("The solution")
+            .build()
+        def withoutSolution = createProblemBuilder()
+            .id("id", "display-name", level1Group)
+            .contextualLabel("Without solution")
+            .details("Just details")
+            .build()
+
+        when:
+        problemWriter.write([withSolution, withoutSolution], writer)
+
+        then:
+        renderedProblem == denormalizeAndStrip('''
+display-name
+  With solution
+    Possible solution: The solution.
+display-name
+  Without solution
+    Just details
+        ''')
+    }
+
     @Issue("https://github.com/gradle/gradle/issues/32016")
     def "java compilation reports are properly separated"() {
         given:

--- a/platforms/ide/problems-rendering/src/test/groovy/org/gradle/problems/internal/rendering/SimpleProblemWriterTest.groovy
+++ b/platforms/ide/problems-rendering/src/test/groovy/org/gradle/problems/internal/rendering/SimpleProblemWriterTest.groovy
@@ -163,8 +163,8 @@ Problem found: Project is a prototype (id: sample-problems:prototype-project)
 Problem found: Project is a prototype (id: sample-problems:prototype-project)
   This is a prototype and not a guideline for modeling real-life projects
     Complex build logic like the Problems API usage should integrated into plugins
-    Solution: Look up the samples index for real-life examples
     Location: /path/to/script line 20
+    Possible solution: Look up the samples index for real-life examples.
         ''')
     }
 
@@ -189,10 +189,11 @@ Problem found: Project is a prototype (id: sample-problems:prototype-project)
 Problem found: Project is a prototype (id: sample-problems:prototype-project)
   This is a prototype and not a guideline for modeling real-life projects
     Complex build logic like the Problems API usage should integrated into plugins
-    Solution: Look up the samples index for real-life examples
-    Solution: Or read the documentation on the Gradle website
     Location: /path/to/script line 20
     Location: /path/to/alternative line 30
+    Possible solutions:
+      1. Look up the samples index for real-life examples.
+      2. Or read the documentation on the Gradle website.
         ''')
     }
 
@@ -213,8 +214,51 @@ Problem found: Project is a prototype (id: sample-problems:prototype-project)
   This is a prototype and not a guideline for modeling real-life projects
     Complex build logic like the Problems API
     usage should integrated into plugins
-    Solution: Look up the samples index for
-              real-life examples
+    Possible solution: Look up the samples index for
+                       real-life examples.
+        ''')
+    }
+
+    def "render multiline messages with multiple solutions"() {
+        def problem = createProblem { ProblemBuilderInternal spec ->
+            spec.id(createId())
+                .contextualLabel("This is a prototype and not${System.lineSeparator()}a guideline for modeling real-life projects") // API enforces a single line for contextual label
+                .details("Complex build logic like the Problems API${System.lineSeparator()}usage should integrated into plugins")
+                .solution("Look up the samples index for${System.lineSeparator()}real-life examples")
+                .solution("Another interesting${System.lineSeparator()}alternative solution")
+                .solution("3rd solution")
+                .solution("4th solution")
+                .solution("5th solution")
+                .solution("6th solution")
+                .solution("7th solution")
+                .solution("8th solution")
+                .solution("9th solution")
+                .solution("10th solution${System.lineSeparator()}with some content")
+        }
+
+        when:
+        problemWriter.write(problem, writer)
+
+        then:
+        renderedProblem == denormalizeAndStrip('''
+Problem found: Project is a prototype (id: sample-problems:prototype-project)
+  This is a prototype and not a guideline for modeling real-life projects
+    Complex build logic like the Problems API
+    usage should integrated into plugins
+    Possible solutions:
+      1. Look up the samples index for
+         real-life examples.
+      2. Another interesting
+         alternative solution.
+      3. 3rd solution.
+      4. 4th solution.
+      5. 5th solution.
+      6. 6th solution.
+      7. 7th solution.
+      8. 8th solution.
+      9. 9th solution.
+      10. 10th solution
+          with some content.
         ''')
     }
 

--- a/platforms/native/platform-native/src/integTest/groovy/org/gradle/nativeplatform/NativeDependentComponentsReportIntegrationTest.groovy
+++ b/platforms/native/platform-native/src/integTest/groovy/org/gradle/nativeplatform/NativeDependentComponentsReportIntegrationTest.groovy
@@ -226,10 +226,13 @@ class NativeDependentComponentsReportIntegrationTest extends AbstractIntegration
     }
 
     String removeIrrelevantOutput(String output) {
-        return output.readLines().findAll {
-             !(it ==~ /^Problem found.*$/) && !(it ==~ /.*caused invocation of 'Task.project' in other task at execution.*$/)
-            && !(it ==~ /.*Documentation: https:\/\/docs.gradle.org\/.*$/)
-        }.join('\n')
+        // Strip entire "Problem found: ... For more information, ..." blocks that bleed into this
+        // report from configuration-cache problem rendering. The "For more information" line ends
+        // with either a URL, a URL + ".", or "... in the Gradle documentation." depending on the
+        // DocLink implementation, so match greedily to end-of-line.
+        String stripped = output.replaceAll(/(?s)Problem found:.*?For more information, please refer to [^\n]*/, '')
+        // Collapse runs of blank lines (left behind by the stripped blocks) down to one.
+        return stripped.replaceAll(/\n\s*\n(\s*\n)+/, '\n\n')
     }
 
     @Requires(TestExecutionPreconditions.NotParallelExecutor)

--- a/subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java
@@ -423,13 +423,12 @@ public class BuildExceptionReporter implements Action<Throwable> {
 
         Collection<ProblemInternal> all = failure.getProblems();
         for (ProblemInternal problem : all) {
-            // TODO (donat) integrate type validation failure rendering with problems-rendering and get rid of the special-casing here
+            // Java compilation problems are rendered by JavaCompilationWriter which emits only the details,
+            // so their solutions still need to reach the user via the resolution section.
+            // All other problem writers render solutions and doc links inline.
             ProblemGroup group = problem.getDefinition().getId().getGroup();
-            if (!GradleCoreProblemGroup.validation().type().equals(group) && !GradleCoreProblemGroup.validation().property().equals(group)) {
+            if (GradleCoreProblemGroup.compilation().java().equals(group)) {
                 resolutions.addAll(problem.getSolutions());
-                if (problem.getDefinition().getDocumentationLink() != null) {
-                    resolutions.add("For more information, see " + problem.getDefinition().getDocumentationLink().getUrl());
-                }
             }
         }
 

